avm/res/insights/data-collection-rule/main.bicep

```diff
@@ -75,78 +70,61 @@ resource avmTelemetry 'Microsoft.Resources/deployments@2024-03-01' = if (enableT
   }
 }
 
-resource dataCollectionRule 'Microsoft.Insights/dataCollectionRules@2023-03-11' = {
+resource dataCollectionRule 'Microsoft.Insights/dataCollectionRules@2023-03-11' = if (dataCollectionRuleProperties.kind != 'All') {
   kind: dataCollectionRuleProperties.kind
   location: location
   name: name
   tags: tags
-  properties: union(
-    {
-      description: dataCollectionRuleProperties.?description
-    },
-    dataCollectionRuleProperties.kind == 'Linux' || dataCollectionRuleProperties.kind == 'Windows'
-      ? {
-          dataSources: dataCollectionRuleProperties.dataSources
-          dataFlows: dataCollectionRuleProperties.dataFlows
-          destinations: dataCollectionRuleProperties.destinations
-          dataCollectionEndpointId: dataCollectionRuleProperties.?dataCollectionEndpointResourceId
-          streamDeclarations: dataCollectionRuleProperties.?streamDeclarations
-        }
-      : {},
-    dataCollectionRuleProperties.kind == 'AgentSettings'
-      ? {
-          agentSettings: dataCollectionRuleProperties.agentSettings
-        }
-      : {}
-  )
+  properties: dataCollectionRulePropertiesUnion
 }
 
-resource dataCollectionRule_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock ?? {}) && lock.?kind != 'None') {
-  name: lock.?name ?? 'lock-${name}'
-  properties: {
-    level: lock.?kind ?? ''
-    notes: lock.?kind == 'CanNotDelete'
-      ? 'Cannot delete resource or child resources.'
-      : 'Cannot delete or modify the resource or child resources.'
+// Using a separate resource for parameter kind: 'All' as it requires that the "kind' is not set on the resource and 'kind: null' is not allowed for this resource type
+resource dataCollectionRuleAll 'Microsoft.Insights/dataCollectionRules@2023-03-11' = if (dataCollectionRuleProperties.kind == 'All') {
+  location: location
+  name: name
+  tags: tags
+  properties: dataCollectionRulePropertiesUnion
+}
+
+// Using a module as a workaround for issues with conditional scope: https://github.com/Azure/bicep/issues/7367
+module dataCollectionRule_lock 'modules/nested_lock.bicep' = if (!empty(lock ?? {}) && lock.?kind != 'None') {
```

Comment:
I'm sure this works - but - I can't help but wonder if we should not have both the lock & roleAssignment in one nested file. This not only simplifies this template, but also means 1 less deployment.